hail/src/main/scala/is/hail/expr/ir/agg/ArrayElementRegionValueAggregator.scala

```diff
@@ -0,0 +1,269 @@
+                s.copyFrom(eltState(sab.arrayIdx, i))
+              }),
+            container.addState(sab),
+            sab.advance()))),
+      off := srvb.end())
+  }
+}
+
+class ArrayElementRegionValueAggregator(nestedAggs: Array[StagedRegionValueAggregator], knownLength: Boolean) extends StagedRegionValueAggregator {
+  type State = ArrayElementState
+  private val nStates: Int = nestedAggs.length
+
+  var initOpTypes: Array[PType] = nestedAggs.flatMap(_.initOpTypes)
+  if (knownLength)
+    initOpTypes = PInt32() +: initOpTypes
+  val seqOpTypes: Array[PType] = Array(PInt32())
+
+  val resultEltType: PTuple = PTuple(nestedAggs.map(_.resultType))
+  val resultType: PArray = PArray(resultEltType)
+
+  def createState(mb: EmitMethodBuilder): State = ArrayElementState.create(mb, nestedAggs, knownLength)
+
+  // inits all things
+  def initOp(state: State, init: Array[RVAVariable], dummy: Boolean): Code[Unit] = {
+    var i = if (knownLength) 1 else 0
+    val initOps = Array.tabulate(nStates) { sIdx =>
+      val agg = nestedAggs(sIdx)
+      val vars = init.slice(i, i + agg.initOpTypes.length)
+      i += agg.initOpTypes.length
+      () => agg.initOp(state.nested(sIdx), vars)
+    }
+
+    if (knownLength) {
+      val len = init.head
+      assert(len.t isOfType PInt32())
+      Code(state.init(initOps), len.setup,
+        state.initLength(len.m.mux(Code._fatal("Array length can't be missing"), len.v[Int])))
+    } else {
+      Code(state.init(initOps), state.lenRef := -1)
+    }
+  }
+
+  //does a length check on arrays
+  def seqOp(state: State, seq: Array[RVAVariable], dummy: Boolean): Code[Unit] = {
+    val Array(len) = seq
+    assert(len.t isOfType PInt32())
+
+    Code(len.setup, len.m.mux(Code._empty, state.checkLength(len.v[Int])))
+  }
+
+  def combOp(state: State, other: State, dummy: Boolean): Code[Unit] = {
+    Code(
+      (other.lenRef < 0).mux(
+        (state.lenRef < 0).mux(
+          Code._empty,
+          other.initLength(state.lenRef)),
+        state.checkLength(other.lenRef)),
+      state.idx := 0,
+      Code.whileLoop(state.idx < state.lenRef,
+        other.scoped(state.idx,
+          state.update(state.idx)( (i, s) =>
+            nestedAggs(i).combOp(s, other.nested(i)))),
+        state.idx := state.idx + 1))
+  }
+
+  def result(state: State, srvb: StagedRegionValueBuilder, dummy: Boolean): Code[Unit] =
+    srvb.addArray(resultType, { sab =>
+      Code(
+        sab.start(state.lenRef),
+        Code.whileLoop(sab.arrayIdx < state.lenRef,
+          sab.addBaseStruct(resultEltType, { ssb =>
+            Code(
+              ssb.start(),
+              state.scoped(sab.arrayIdx) { (i, s) =>
+                Code(nestedAggs(i).result(s, ssb), ssb.advance())
+              })
+          }),
+          sab.advance()))
+    })
+}
+
+class ArrayElementRegionValueAggregator2(nestedAggs: Array[StagedRegionValueAggregator]) extends StagedRegionValueAggregator {
```

Comment:
What does this do? Is it supposed to have unimplemented methods?

Reply:
(for the sake of logging this conversation; this preserves the same structure as the old ArrayElementsAggregator, and I'll rename these accordingly.

The first one contains an initOp that initializes the state, and a seqOp that takes the array length and checks it.

The second one doesn't have an initOp; its seqOp takes two arguments: the index of the agg state array to update, and a second argument of type TVoid that defines the seqOp to perform on it. Since the way it's extracted enforces that the seq op is always structured like:
```
Begin(
  SeqOp(i, Seq(len), agg1),
  ArrayFor(array, elt,
    SeqOp(i, Seq(idx, nestedSeqOp), agg2)))
```

it should never need to construct a state. I'll rename the classes and errors to be more descriptive.